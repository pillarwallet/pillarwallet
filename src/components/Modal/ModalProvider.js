// @flow
/*
    Pillar Wallet: the personal data locker
    Copyright (C) 2019 Stiftung Pillar Project

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License along
    with this program; if not, write to the Free Software Foundation, Inc.,
    51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
*/

import React, { useContext } from 'react';
import type { Node as ReactNode } from 'react';
import { Keyboard } from 'react-native';

export type ModalOptions = {|
  render: () => ReactNode,
|}

type ModalState = {|
  ...ModalOptions,
  id: string,
  openedAt: Date,
|};

// `pending` is null, except for the case when ModalProvider is in process of
// closing all visible modals. While that happens, other open/close operations
// should be performed on the `pending` array. Afterwards, all modals from
// `pending` are moved to `stack` to be rendered.

type State = {|
  stack: ModalState[],
  pending: null | ModalState[],
|};

type ModalInstance = {
  closeRNModal: () => Promise<void>,
};

const ModalStackContext = React.createContext<ModalState[]>([]);
const ModalNextIndexContext = React.createContext<number>(0);

export const EMPTY_MODAL_ID = 'default value for modal id context';
export const ModalIdContext = React.createContext<string>(EMPTY_MODAL_ID);

export const ModalStack = () => {
  const stack = useContext(ModalStackContext);
  const index = useContext(ModalNextIndexContext);
  const options = stack[index];

  return (index < stack.length ? (
    <ModalNextIndexContext.Provider value={index + 1}>
      <ModalIdContext.Provider value={options.id}>
        {options.render()}
      </ModalIdContext.Provider>
    </ModalNextIndexContext.Provider>
  ) : null);
};

class ModalProvider extends React.Component<{||}, State> {
  static _activeInstance: ModalProvider | null = null;
  static getInstance(): ModalProvider | null {
    return this._activeInstance;
  }

  state: State = {
    stack: [],
    pending: null,
  }

  _idCounter = 0;
  modalInstances: Map<string, ModalInstance> = new Map();

  componentDidMount() {
    if (ModalProvider._activeInstance === null) {
      ModalProvider._activeInstance = this;
    }
  }

  componentWillUnmount() {
    if (ModalProvider._activeInstance === this) {
      ModalProvider._activeInstance = null;
    }
  }

  open: (options: ModalOptions) => void = (options: ModalOptions) => {
    const id = (this._idCounter++).toString();
    const fullOptions = {
      ...options,
      id,
      openedAt: new Date(),
    };

    this.setState(({ stack, pending }) =>
      pending === null
        ? { stack: [...stack, fullOptions] }
        : { pending: [...pending, fullOptions] },
    );
  }

  close: (id: string) => void = (id: string) => {
    const matchModal = ({ id: _id }) => _id === id;

    if (this.state.pending !== null) {
      this.setState(({ pending }) => ({
        pending: pending && pending.filter(m => !matchModal(m)),
      }));

      return;
    }

    const index = this.state.stack.findIndex(matchModal);
    const instance = this.modalInstances.get(id);
    if (!instance || index === -1) return;

    (async () => {
      Keyboard.dismiss();

      // If this is the modal on top of the stack, close with animation before
      // removing from state. The other case shouldn't happen, so while we make
      // sure the proper element was removed from stack array, there is no
      // special handling.
      if (index === this.state.stack.length - 1) {
        await instance.closeRNModal();
      }

      this.setState(({ stack }) => ({
        stack: stack.filter(m => !matchModal(m)),
      }));
    })();
  }

  closeAll = () => {
    const now = new Date();

    // In case closeAll was called because of a navigation event, it might have
    // happened after the destination screen was rendered with a different modal,
    // so modals opened right before this moment are ignored.
    const wasJustOpened = ({ openedAt }) => now - openedAt < 100;

    this.setState(({ stack, pending }) => pending === null ? ({
      stack: stack.filter(m => !wasJustOpened(m)),
      pending: stack.filter(wasJustOpened),
    }) : {
      stack,
      pending: pending.filter(wasJustOpened),
    }, async () => {
      await Promise.all(this.state.stack.map(({ id }) => {
        const modal = this.modalInstances.get(id);
        return modal && modal.closeRNModal();
      }));

      this.setState(({ stack, pending }) => ({
        stack: pending ?? stack,
        pending: null,
      }));
    });
  }

  render() {
    return (
      <ModalStackContext.Provider value={this.state.stack}>
        <ModalNextIndexContext.Provider value={0}>
          <ModalStack />
        </ModalNextIndexContext.Provider>
      </ModalStackContext.Provider>
    );
  }
}

export default ModalProvider;
