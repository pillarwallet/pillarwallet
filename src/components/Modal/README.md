# Modals

## Basic usage: `Modal.open()`

To open a modal, call `Modal.open(render)`, where `render` is a function that
returns a `React.Node`.

The root component of the result should be `<Modal>` imported from `components/Modal`.
It can be indirect. For example, if `<BoxModal>` is defined as

```js
const BoxModal = () => (
  <Modal> ... </Modal>
);
```

then calling `Modal.open(() => <BoxModal />)` will work as intended.

## `Modal.closeAll()`

Will close visible modals and queue any new ones to display after the animation
completes.

## Creating new modals

The component from `./Modal` should be used as the main wrapper for any modal.
It renders `react-native-modal` and passes through most of its props, but does
not take `isVisible`. The assumption is that modals become visible when they
get included in the component tree, on calling `Modal.open`, and stay that way
until they're closed and removed from the stack.

It's important to note that props passed in the render function given
to `Modal.open` won't be updated on later re-renders of the parent component.
If the content inside the modal should update in response to Redux actions, the
modal component itself should be wrapped with `connect`. Likewise, if there are
stateful elements in the modal (e.g. toggles, text inputs) then this state
should also be defined in the modal, not the parent.

`<Modal>` instance can be closed by calling `.close()` on a ref passed to it.
Expanding the example above, a modal that closes itself on a button press would
look like this:

```js
const BoxModal = () => {
  const ref = useRef();

  return (
    <Modal ref={ref}>
      <Button onPress={() => ref.current?.close()} />
    </Modal>
  );
};
```

## Implementation

`ModalProvider` is another part aside from the `Modal` class that makes modals
work. Modals are organized in a stack. `ModalProvider` is responsible for
keeping track of modals on the stack and rendering them, which is why it's
necessary for it to be included somewhere in the application.

There should be only one instance of `<ModalProvider>` rendered at a time. It's
not possible to show modals in two separate places simultaneously, so any
additional ones are ignored.

The "public" `modal.close` function doesn't close the modal directly - instead
it's redirected to the `<ModalProvider>` instance, which updates the stack and
calls the inner one (`modal.closeRNModal`).

In order to display multiple modals one over another, each modal on the stack
needs to be rendered as a child of the one before it. At the same time, any
component that wants to render as a modal should have the ability to set the
props of `<Modal>` directly to have control over layout and scrollable content.

To deal with those constraints, `<ModalProvider>` passes the needed extra data
via context. It's divided into three parts:

- `ModalStackContext`: an array with all open modals
- `ModalNextIndexContext`: where in the stack is the next modal to render
- `ModalIdContext`: the id of the current modal (when asking to be closed,
  a modal needs a way to identify itself to its `ModalProvider`)

`ModalStackContext` is set at the root of `<ModalProvider>` and stays the same
for all modals.

The `<ModalStack>` helper renders the "rest" of the stack and advances the
index kept in `ModalNextIndexContext`. It's included in `<ModalProvider>` to
start rendering modals and in each `<Modal>` to enable the recursion.

Some of the complexity of `ModalProvider`'s state is due to the order of
navigation event notification and screen changes, which would otherwise cause
problems with modals opened on entering a screen (e.g. the introduction overlay
on the Exchange screen).
