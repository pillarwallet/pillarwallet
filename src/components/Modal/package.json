{
  "name": "Modal.js",
  "version": "0.0.0",
  "main": "./Modal.js"
}
