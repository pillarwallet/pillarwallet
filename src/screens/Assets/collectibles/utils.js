// @flow
/*
    Pillar Wallet: the personal data locker
    Copyright (C) 2021 Stiftung Pillar Project

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License along
    with this program; if not, write to the Free Software Foundation, Inc.,
    51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
*/

// constants
import { COLLECTIBLES } from 'constants/assetsConstants';

// types
import type { Collectible } from 'models/Collectible';
import type { Chain } from 'models/Chain';
import type { CollectibleItem } from './selectors';

export function buildCollectibleFromCollectibleItem(
  collectibleItem: CollectibleItem,
  chain: Chain,
): Collectible {
  const {
    id,
    title: name,
    description,
    imageUrl,
    iconUrl,
    contractAddress,
    tokenId,
  } = collectibleItem;

  return {
    id,
    name,
    description,
    image: imageUrl,
    icon: iconUrl,
    imageUrl,
    iconUrl,
    contractAddress,
    tokenId,
    chain,
    tokenType: COLLECTIBLES,
  };
}
