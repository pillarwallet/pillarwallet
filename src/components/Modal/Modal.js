// @flow
/*
    Pillar Wallet: the personal data locker
    Copyright (C) 2019 Stiftung Pillar Project

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License along
    with this program; if not, write to the Free Software Foundation, Inc.,
    51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
*/

import React from 'react';
import type { Node as ReactNode } from 'react';
import RNModal from 'react-native-modal';
import { View } from 'react-native';
import styled from 'styled-components/native';

import { ToastProvider } from 'components/Toast';

import ModalProvider, { ModalStack, ModalIdContext, EMPTY_MODAL_ID } from './ModalProvider';
import type { ModalOptions } from './ModalProvider';

export { default as ModalProvider } from './ModalProvider';

export type ScrollToProps = {|
  x?: number,
  y?: number,
  animated?: boolean,
|}

type SwipeDirection = 'up' | 'down' | 'left' | 'right';

// Props according to react-native-modal
type Props = {|
  children?: ReactNode,
  animationIn?: string | Object,
  animationInTiming?: number,
  animationOut?: string | Object,
  animationOutTiming?: number,
  avoidKeyboard?: boolean,
  coverScreen?: boolean,
  hasBackdrop?: boolean,
  backdropColor?: string,
  backdropOpacity?: number,
  backdropTransitionInTiming?: number,
  backdropTransitionOutTiming?: number,
  customBackdrop?: ReactNode,
  deviceHeight?: number,
  deviceWidth?: number,
  onModalWillHide?: () => void,
  onModalHide?: () => void,
  onModalWillShow?: () => void,
  onModalShow?: () => void,
  onSwipeStart?: () => void,
  onSwipeMove?: (percentageShown: number) => void,
  scrollOffset?: number,
  scrollOffsetMax?: number,
  scrollTo?: (_: ScrollToProps) => void,
  scrollHorizontal?: boolean,
  swipeThreshold?: number,
  swipeDirection?: SwipeDirection | SwipeDirection[],
  useNativeDriver?: boolean,
  hideModalContentWhileAnimating?: boolean,
  style?: any,
|};

type State = {|
  isVisible: boolean,
|};

const StaticContainer = styled.View`
  background-color: rgba(0, 0, 0, 0.7);
  flex: 1;
  justify-content: center;
`;

class Modal extends React.Component<Props, State> {
  static open(render: $PropertyType<ModalOptions, 'render'>) {
    const instance = ModalProvider.getInstance();
    if (instance) instance.open({ render });
  }

  static closeAll() {
    const instance = ModalProvider.getInstance();
    if (instance) instance.closeAll();
  }

  static contextType = ModalIdContext;

  // alias for readability
  getId = () => this.context

  state = {
    isVisible: true,
  }

  closingCallbacks: (() => void)[] = [];

  callClosingCallbacks = () => {
    this.closingCallbacks.forEach(callback => callback());
    this.closingCallbacks = [];
  }

  componentDidMount() {
    const instance = ModalProvider.getInstance();
    if (instance) instance.modalInstances.set(this.getId(), this);
  }

  componentWillUnmount() {
    // Don't let the closeRNModal method hang if RNModal's onModalHide doesn't get
    // called before unmounting.
    this.callClosingCallbacks();

    const instance = ModalProvider.getInstance();
    if (instance) instance.modalInstances.delete(this.getId());
  }

  close = () => {
    const instance = ModalProvider.getInstance();
    if (instance) instance.close(this.getId());
  }

  closeRNModal = () => {
    this.setState({ isVisible: false });
    return new Promise((resolve) => {
      this.closingCallbacks.push(resolve);
    });
  }

  onHide = () => {
    this.callClosingCallbacks();
    const { onModalHide } = this.props;
    if (onModalHide) onModalHide();
  }

  render() {
    // swipeDirection is passed explicitly to work around issues with using
    // spread operator on union types
    const { children, swipeDirection, ...props } = this.props;
    const { isVisible } = this.state;

    // render contents direrctly with overlay-like background if the modal was
    // attached outside of ModalProvider (e.g. in storybook)
    return this.getId() === EMPTY_MODAL_ID ? (
      <StaticContainer>
        <View style={this.props.style}>
          {children}
        </View>
      </StaticContainer>
    ) : (
      <RNModal
        isVisible={isVisible}
        propagateSwipe
        swipeDirection={swipeDirection ?? 'down'}
        onSwipeComplete={this.close}
        onBackdropPress={this.close}
        onBackButtonPress={this.close}
        avoidKeyboard
        style={{ margin: 0 }}
        {...props}
        onModalHide={this.onHide}
      >
        {children}
        {isVisible && <ToastProvider />}
        <ModalStack />
      </RNModal>
    );
  }
}

export default Modal;
