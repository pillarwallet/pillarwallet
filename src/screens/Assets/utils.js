// @flow
/*
    Pillar Wallet: the personal data locker
    Copyright (C) 2019 Stiftung Pillar Project

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License along
    with this program; if not, write to the Free Software Foundation, Inc.,
    51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
*/

import * as React from 'react';
import { Platform, LayoutAnimation } from 'react-native';

// Configs
import { getEnv } from 'configs/envConfig';

// Utils
import { LIST_ITEMS_APPEARANCE } from 'utils/layoutAnimations';

// Types
import type { Asset } from 'models/Asset';
import type { Chain } from 'models/Chain';

type AssetDetailsContext = {|
  accountAddress: ?string,
|};

/**
 * Extracted from AssetList.js. Asset screen expects specific but untyped format of data as navigation param.
 */
export function buildAssetDataNavigationParam(asset: Asset, { accountAddress }: AssetDetailsContext) {
  const { symbol, name, iconUrl, decimals, iconMonoUrl, patternUrl, wallpaperUrl } = asset;

  const fullIconMonoUrl = iconMonoUrl ? `${getEnv().SDK_PROVIDER}/${iconMonoUrl}?size=2` : '';
  const fullIconWallpaperUrl = `${getEnv().SDK_PROVIDER}/${wallpaperUrl}${Platform.OS === 'ios' ? '?size=3' : ''}`;
  const fullIconUrl = iconUrl ? `${getEnv().SDK_PROVIDER}/${iconUrl}?size=3` : '';
  const patternIcon = patternUrl ? `${getEnv().SDK_PROVIDER}/${patternUrl}?size=3` : fullIconUrl;

  return {
    id: symbol,
    name: name || symbol,
    token: symbol,
    address: accountAddress,
    contractAddress: asset.address,
    icon: fullIconMonoUrl,
    wallpaper: fullIconWallpaperUrl,
    iconColor: fullIconUrl,
    patternIcon,
    description: asset.description,
    decimals,
  };
}

export type FlagPerChain = { [Chain]: ?boolean };

export function useExpandItemsPerChain(initialChain: ?Chain) {
  const [expandItemsPerChain, setExpandItemsPerChain] = React.useState<FlagPerChain>(
    // $FlowFixMe: type inference limitation
    initialChain ? { [initialChain]: true } : {},
  );

  const toggleExpandItems = (chain: Chain) => {
    LayoutAnimation.configureNext(LIST_ITEMS_APPEARANCE);
    // $FlowFixMe: type inference limitation
    setExpandItemsPerChain({ ...expandItemsPerChain, [chain]: !expandItemsPerChain[chain] });
  };

  return { expandItemsPerChain, toggleExpandItems };
}
