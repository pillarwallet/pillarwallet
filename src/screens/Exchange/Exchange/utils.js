// @flow
/*
    Pillar Wallet: the personal data locker
    Copyright (C) 2019 Stiftung Pillar Project

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License along
    with this program; if not, write to the Free Software Foundation, Inc.,
    51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
*/

import { BigNumber } from 'bignumber.js';

// Constants
import { ETH } from 'constants/assetsConstants';

// Utils
import { getBalance, sortAssets, getAssetOption } from 'utils/assets';

// Types
import type { Rates, Asset, Assets, AssetOption, Currency } from 'models/Asset';
import type { WalletAssetsBalances } from 'models/Balances';

export const getExchangeFromAssetOptions = (
  assets: Assets,
  supportedAssets: Asset[],
  balances: WalletAssetsBalances,
  currency: Currency,
  rates: Rates,
): AssetOption[] => {
  const isMatching = (asset: Asset) => asset.symbol === ETH || getBalance(balances, asset.symbol) !== 0;
  const isSupported = (asset: Asset) =>
    supportedAssets.some((supportedAsset) => asset.symbol === supportedAsset.symbol);

  return sortAssets(assets)
    .filter((asset) => isMatching(asset) && isSupported(asset))
    .map((asset) => getAssetOption(asset, balances, rates, currency));
};

export const getExchangeToAssetOptions = (
  supportedAssets: Asset[],
  balances: WalletAssetsBalances,
  currency: Currency,
  rates: Rates,
): AssetOption[] => {
  if (!Array.isArray(supportedAssets)) return [];

  return supportedAssets.map((asset) => getAssetOption(asset, balances, rates, currency));
};

export const shouldTriggerSearch = (
  fromAsset: ?AssetOption,
  toAsset: ?AssetOption,
  fromAmount: string,
): boolean %checks => {
  return (
    !!fromAsset &&
    !!toAsset &&
    !!+fromAmount &&
    fromAmount[fromAmount.length - 1] !== '.' &&
    fromAsset.symbol !== toAsset.symbol &&
    isEnoughAssetBalance(fromAsset.assetBalance, fromAmount)
  );
};

const isEnoughAssetBalance = (assetBalance: ?string, amount: string): boolean => {
  try {
    const amountBN = new BigNumber(amount);
    const balanceBN = new BigNumber(assetBalance ?? 0);
    // assetBalance is fixed to 6 digits and amount is not, so usually amount will be technically higher
    // fix and round both down to 6 to get meaningful info
    const amountFixed = amountBN.toFixed(6, 1);
    const balanceFixed = balanceBN.toFixed(6, 1);
    return new BigNumber(balanceFixed).isGreaterThanOrEqualTo(new BigNumber(amountFixed));
  } catch {
    return false;
  }
};
