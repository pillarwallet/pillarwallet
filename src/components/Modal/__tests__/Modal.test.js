// @flow
/*
    Pillar Wallet: the personal data locker
    Copyright (C) 2019 Stiftung Pillar Project

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License along
    with this program; if not, write to the Free Software Foundation, Inc.,
    51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
*/

import React from 'react';
import { act, create } from 'react-test-renderer';
import { ThemeProvider } from 'styled-components/native';

import { defaultTheme } from 'utils/themes';

import Modal, { ModalProvider } from '../Modal';

describe('Modal', () => {
  let rendererInstance: $FlowFixMe = null;

  beforeEach(() => {
    jest.useFakeTimers();

    act(() => {
      rendererInstance = create(
        <ThemeProvider theme={defaultTheme}>
          <ModalProvider />
        </ThemeProvider>,
      );
    });

    jest.runAllTimers();
  });

  afterEach(() => {
    act(() => rendererInstance.unmount());
    jest.useRealTimers();
  });

  it('opens a modal', () => {
    Modal.open(() => <Modal />);
    jest.runAllTimers();
    expect(rendererInstance.toJSON()).toMatchSnapshot();
  });

  it('opens multiple modals', () => {
    Modal.open(() => <Modal />);
    Modal.open(() => <Modal />);
    jest.runAllTimers();
    expect(rendererInstance.toJSON()).toMatchSnapshot();
  });
});
