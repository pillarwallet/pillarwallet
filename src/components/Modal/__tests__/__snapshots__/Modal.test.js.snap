// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal opens a modal 1`] = `
<Modal
  animationType="none"
  hardwareAccelerated={false}
  hideModalContentWhileAnimating={false}
  onModalHide={[Function]}
  onModalWillHide={[Function]}
  onModalWillShow={[Function]}
  onRequestClose={[Function]}
  onSwipeComplete={[Function]}
  scrollHorizontal={false}
  scrollOffset={0}
  scrollOffsetMax={0}
  scrollTo={null}
  supportedOrientations={
    Array [
      "portrait",
      "landscape",
    ]
  }
  swipeDirection="down"
  swipeThreshold={100}
  transparent={true}
  visible={true}
>
  <View
    accessible={true}
    focusable={true}
    forwardedRef={[Function]}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "backgroundColor": "black",
        "bottom": 0,
        "height": 1334,
        "left": 0,
        "opacity": 0.7,
        "position": "absolute",
        "right": 0,
        "top": 0,
        "width": 750,
      }
    }
  />
  <View
    onLayout={[Function]}
    pointerEvents="box-none"
    style={
      Array [
        Array [
          Object {
            "margin": 37.5,
            "transform": Array [
              Object {
                "translateY": 0,
              },
            ],
          },
          Object {
            "flex": 1,
            "justifyContent": "center",
          },
          Object {
            "margin": 0,
          },
          Object {
            "margin": 0,
          },
        ],
        Object {
          "paddingBottom": 0,
        },
      ]
    }
  >
    <View
      forwardedRef={[Function]}
      hideModalContentWhileAnimating={false}
      onModalHide={[Function]}
      onModalWillHide={[Function]}
      onModalWillShow={[Function]}
      onMoveShouldSetResponder={[Function]}
      onMoveShouldSetResponderCapture={[Function]}
      onResponderEnd={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderReject={[Function]}
      onResponderRelease={[Function]}
      onResponderStart={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      onStartShouldSetResponderCapture={[Function]}
      onSwipeComplete={[Function]}
      pointerEvents="box-none"
      scrollHorizontal={false}
      scrollOffset={0}
      scrollOffsetMax={0}
      scrollTo={null}
      style={
        Object {
          "flex": 1,
          "justifyContent": "center",
          "left": 0,
          "margin": 0,
          "top": 0,
          "transform": Array [
            Object {
              "translateY": 0,
            },
          ],
        }
      }
      supportedOrientations={
        Array [
          "portrait",
          "landscape",
        ]
      }
      swipeDirection="down"
      swipeThreshold={100}
    >
      <View
        forwardedRef={[Function]}
        onLayout={[Function]}
        pointerEvents="box-none"
        statusBarHeight={null}
        style={
          Object {
            "left": 0,
            "marginTop": 40,
            "paddingBottom": 0,
            "paddingLeft": 20,
            "paddingRight": 20,
            "paddingTop": 20,
            "position": "absolute",
            "top": 0,
            "width": "100%",
            "zIndex": 1000,
          }
        }
      >
        <View
          style={
            Object {
              "height": 0,
            }
          }
        />
      </View>
    </View>
  </View>
</Modal>
`;

exports[`Modal opens multiple modals 1`] = `
<Modal
  animationType="none"
  hardwareAccelerated={false}
  hideModalContentWhileAnimating={false}
  onModalHide={[Function]}
  onModalWillHide={[Function]}
  onModalWillShow={[Function]}
  onRequestClose={[Function]}
  onSwipeComplete={[Function]}
  scrollHorizontal={false}
  scrollOffset={0}
  scrollOffsetMax={0}
  scrollTo={null}
  supportedOrientations={
    Array [
      "portrait",
      "landscape",
    ]
  }
  swipeDirection="down"
  swipeThreshold={100}
  transparent={true}
  visible={true}
>
  <View
    accessible={true}
    focusable={true}
    forwardedRef={[Function]}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "backgroundColor": "black",
        "bottom": 0,
        "height": 1334,
        "left": 0,
        "opacity": 0.7,
        "position": "absolute",
        "right": 0,
        "top": 0,
        "width": 750,
      }
    }
  />
  <View
    onLayout={[Function]}
    pointerEvents="box-none"
    style={
      Array [
        Array [
          Object {
            "margin": 37.5,
            "transform": Array [
              Object {
                "translateY": 0,
              },
            ],
          },
          Object {
            "flex": 1,
            "justifyContent": "center",
          },
          Object {
            "margin": 0,
          },
          Object {
            "margin": 0,
          },
        ],
        Object {
          "paddingBottom": 0,
        },
      ]
    }
  >
    <View
      forwardedRef={[Function]}
      hideModalContentWhileAnimating={false}
      onModalHide={[Function]}
      onModalWillHide={[Function]}
      onModalWillShow={[Function]}
      onMoveShouldSetResponder={[Function]}
      onMoveShouldSetResponderCapture={[Function]}
      onResponderEnd={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderReject={[Function]}
      onResponderRelease={[Function]}
      onResponderStart={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      onStartShouldSetResponderCapture={[Function]}
      onSwipeComplete={[Function]}
      pointerEvents="box-none"
      scrollHorizontal={false}
      scrollOffset={0}
      scrollOffsetMax={0}
      scrollTo={null}
      style={
        Object {
          "flex": 1,
          "justifyContent": "center",
          "left": 0,
          "margin": 0,
          "top": 0,
          "transform": Array [
            Object {
              "translateY": 0,
            },
          ],
        }
      }
      supportedOrientations={
        Array [
          "portrait",
          "landscape",
        ]
      }
      swipeDirection="down"
      swipeThreshold={100}
    >
      <View
        forwardedRef={[Function]}
        onLayout={[Function]}
        pointerEvents="box-none"
        statusBarHeight={null}
        style={
          Object {
            "left": 0,
            "marginTop": 40,
            "paddingBottom": 0,
            "paddingLeft": 20,
            "paddingRight": 20,
            "paddingTop": 20,
            "position": "absolute",
            "top": 0,
            "width": "100%",
            "zIndex": 1000,
          }
        }
      >
        <View
          style={
            Object {
              "height": 0,
            }
          }
        />
      </View>
      <Modal
        animationType="none"
        hardwareAccelerated={false}
        hideModalContentWhileAnimating={false}
        onModalHide={[Function]}
        onModalWillHide={[Function]}
        onModalWillShow={[Function]}
        onRequestClose={[Function]}
        onSwipeComplete={[Function]}
        scrollHorizontal={false}
        scrollOffset={0}
        scrollOffsetMax={0}
        scrollTo={null}
        supportedOrientations={
          Array [
            "portrait",
            "landscape",
          ]
        }
        swipeDirection="down"
        swipeThreshold={100}
        transparent={true}
        visible={true}
      >
        <View
          accessible={true}
          focusable={true}
          forwardedRef={[Function]}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "backgroundColor": "black",
              "bottom": 0,
              "height": 1334,
              "left": 0,
              "opacity": 0.7,
              "position": "absolute",
              "right": 0,
              "top": 0,
              "width": 750,
            }
          }
        />
        <View
          onLayout={[Function]}
          pointerEvents="box-none"
          style={
            Array [
              Array [
                Object {
                  "margin": 37.5,
                  "transform": Array [
                    Object {
                      "translateY": 0,
                    },
                  ],
                },
                Object {
                  "flex": 1,
                  "justifyContent": "center",
                },
                Object {
                  "margin": 0,
                },
                Object {
                  "margin": 0,
                },
              ],
              Object {
                "paddingBottom": 0,
              },
            ]
          }
        >
          <View
            forwardedRef={[Function]}
            hideModalContentWhileAnimating={false}
            onModalHide={[Function]}
            onModalWillHide={[Function]}
            onModalWillShow={[Function]}
            onMoveShouldSetResponder={[Function]}
            onMoveShouldSetResponderCapture={[Function]}
            onResponderEnd={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderReject={[Function]}
            onResponderRelease={[Function]}
            onResponderStart={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            onStartShouldSetResponderCapture={[Function]}
            onSwipeComplete={[Function]}
            pointerEvents="box-none"
            scrollHorizontal={false}
            scrollOffset={0}
            scrollOffsetMax={0}
            scrollTo={null}
            style={
              Object {
                "flex": 1,
                "justifyContent": "center",
                "left": 0,
                "margin": 0,
                "top": 0,
                "transform": Array [
                  Object {
                    "translateY": 0,
                  },
                ],
              }
            }
            supportedOrientations={
              Array [
                "portrait",
                "landscape",
              ]
            }
            swipeDirection="down"
            swipeThreshold={100}
          >
            <View
              forwardedRef={[Function]}
              onLayout={[Function]}
              pointerEvents="box-none"
              statusBarHeight={null}
              style={
                Object {
                  "left": 0,
                  "marginTop": 40,
                  "paddingBottom": 0,
                  "paddingLeft": 20,
                  "paddingRight": 20,
                  "paddingTop": 20,
                  "position": "absolute",
                  "top": 0,
                  "width": "100%",
                  "zIndex": 1000,
                }
              }
            >
              <View
                style={
                  Object {
                    "height": 0,
                  }
                }
              />
            </View>
          </View>
        </View>
      </Modal>
    </View>
  </View>
</Modal>
`;
